stage1/init/kvm.go

```diff
@@ -113,16 +115,39 @@ func mountSharedVolumes(root string, p *stage1commontypes.Pod, ra *schema.Runtim
 }
 
 func doBindMount(source, destination string, readOnly bool, recursive *bool) error {
-	var flags uintptr = syscall.MS_BIND | syscall.MS_REC
+	var flags uintptr = syscall.MS_BIND
+
 	if readOnly {
 		flags |= syscall.MS_RDONLY
 	}
 
 	// Enable recursive by default and remove it if explicitly requested
-	if recursive != nil && *recursive == false {
-		flags ^= syscall.MS_REC
+	recursiveBool := recursive == nil || *recursive == true
+	if recursiveBool {
+		flags |= syscall.MS_REC
+	}
+
+	if err := syscall.Mount(source, destination, "bind", flags, ""); err != nil {
+		return err
 	}
-	return syscall.Mount(source, destination, "bind", flags, "")
+
+	if readOnly && recursiveBool {
+		// Sub-mounts are still read-write, so find them and remount them read-only
+
+		mnts, err := stage0.GetMountsForPrefix(source + "/")
+		if err != nil {
+			return errwrap.Wrap(fmt.Errorf("error getting mounts under %q from mountinfo", source), err)
+		}
+
+		for _, mnt := range mnts {
+			innerAbsPath := destination + strings.Replace(mnt.MountPoint, source, "", -1)
+			if err := syscall.Mount("", innerAbsPath, "none", syscall.MS_REMOUNT|syscall.MS_RDONLY, ""); err != nil {
+				return err
```

Comment:
I reproduced the EINVAL issue with:
```
$ mkdir /tmp/data
$ mkdir /tmp/data/RO /tmp/data/RW
$ sudo mount --bind -o ro /tmp/data/RO /tmp/data/RO
$ sudo mount --bind /tmp/data/RW /tmp/data/RW
$ sudo mount --bind /tmp/data /tmp/data
$ sudo rkt run --debug  --stage1-path=/usr/bin/stage1-kvm.aci --insecure-options=image,ondisk --volume=foo,readOnly=true,recursive=true,kind=host,source=/tmp/data/ docker://busybox --mount=volume=foo,target=/data/
...
stage1: could not prepare mounts
  └─failed prepare mounts for app "busybox"
    └─failed to prepare mount point
      └─could not bind mount path /data/ (s: /tmp/data/, d: /var/lib/rkt/pods/run/d9dd3a69-dd3a-496a-920e-95bed68ff018/stage1/rootfs/opt/stage2/busybox/rootfs/data)
        └─invalid argument
```

This scenario should be added in the tests as well.